homeassistant/components/caldav/calendar.py

```diff
@@ -193,27 +193,49 @@ def __init__(self, calendar, include_all_day, search):
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     def update(self):
         """Get the latest data."""
+        start_of_today = dt.start_of_local_day()
+        start_of_tomorrow = dt.start_of_local_day() + timedelta(days=1)
+
         # We have to retrieve the results for the whole day as the server
         # won't return events that have already started
-        results = self.calendar.date_search(
-            dt.start_of_local_day(), dt.start_of_local_day() + timedelta(days=1)
-        )
+        results = self.calendar.date_search(start_of_today, start_of_tomorrow)
+
+        # Create new vevents for each recurrence of an event that happens today.
+        # For recurring events, some servers return the original event with recurrence rules
+        # and they would not be properly parsed using their original start/end dates.
+        vevents = [event.instance.vevent for event in results]
+        new_vevents = []
+        for vevent in vevents:
+            for start_dt in vevent.getrruleset() or []:
+                _start_of_today = start_of_today
+                _start_of_tomorrow = start_of_tomorrow
+                if self.is_all_day(vevent):
+                    start_dt = start_dt.date()
+                    _start_of_today = _start_of_today.date()
+                    _start_of_tomorrow = _start_of_tomorrow.date()
+                if _start_of_today <= start_dt < _start_of_tomorrow:
+                    new_vevent = copy.deepcopy(vevent)
```

Comment:
How does this work? It's not a dictionary. How does deepcopy know how to copy this ?

Reply:
Magic? 🧙‍♂️ I just tested `new_vevent = vevent` and that passes, too. So I guess deepcopy was unnecessary.